src/advanced-concepts/system-apis/message-inspection.md

```diff
@@ -1 +1,132 @@
+
+In our example these 'function-variable-accessor' function types are:
+
+```motoko
+() -> ();
+() -> Nat;
+() -> Text;
+```
+
+These functions return the variable arguments that were supplied during the call to the public shared function. Thus, they always have unit argument types and return the variable type of the called function.
+
+For example function `f2` takes a `Nat`. If we wanted to _inspect_ this value, we could call the _associated function_ of variant `#f2`, which is of type `() -> Nat`. This function will provide the actual argument passed in the call to the public shared function.
+
+## Actor with Message Inspection
+
+Lets implement the `inspect` system function for the example [actor](/internet-computer-programming-concepts/actors.html) given above:
+
+```motoko
+{{#include _mo/message-inspection.mo}}
+```
+
+We implemented an actor that _inspects_ any call to its [public shared functions](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) and performs checks on _update_ and _oneway_ functions before they are called.
+
+First, the `Principal` module is imported and the [actor](/internet-computer-programming-concepts/actors.html) is declared with three [public shared functions](/internet-computer-programming-concepts/actors.html#public-shared-functions-in-actors) `f1`, `f2` and `f3`.
+
+Then, we defined `CallArgs` as the object type of the _expected_ argument to the `inspect` system function. We then declare the system function and use `CallArgs` to annotate the `args` argument.
+
+We can now access all the information for any function call inside the `inspect` function through the _chosen_ name `args`. (We could have chosen any other name)
+
+Inside `inspect` we first check whether the `caller` field of `args`, which is a `Principal`, is equal to the anonymous principal. If so, then `inspect` returns `false` and the call is _rejected_. Anonymous principals can't call any functions of this [actor](/internet-computer-programming-concepts/actors.html).
+
+Another check is performed on the size of the `arg` field value of our `args` object. If the _binary message argument_ is larger than `1024` bytes, then the call is _rejected_ by returning `false` once more.
+
+Our `inspect` implementation ends with a switch expression that checks every case of the `msg` variant inside `args`.
+
+In the case function `f1` is called, we ignore the possible arguments supplied to it by using the wildcard `_` for the 'function-variable-accessor' function and always _accept_ the call by always returning `true`.
+
+In case function `f2` is called, we bind the 'function-variable-accessor' function to the local name `f2Args` and run it to get the value of the argument to `f2`, which is a `Nat`. If this value is smaller then `100`, we _accept_ the call, otherwise we _reject_.
```

Comment:
```suggestion
In case function `f2` is called, we bind the 'function-argument-accessor' function to the local name `f2Args` and run it to get the value of the argument to `f2`, which is a `Nat`. If this value is smaller then `100`, we _accept_ the call, otherwise we _reject_.
```